Implement virtio_mmio and pci ecam controller for arm64 kvm plat

justin-he opened the pull request:
To support virtio-* family device, we have 2 choices (virtio mmio, virtio pci)
on arm64 kvm plat. This patch series enable the support of virtio mmio and pci.
virtio pci needs pci controller ECAM mode in arm64 Unikraft.

Test
====
1. The test case is a simple httpreply app.
test qemu cmdline:
a. for virtio mmio
-device virtio-net-device,id=net1,netdev=hostnet1,mac=52:54:00:09:a4:38
b. for virtio pci
-device 
virtio-net-pci,id=net0,netdev=hostnet0,mac=52:54:00:09:a4:31,disable-modern=on

2. Besides, I also tested it with nginx app which was suggested by Felipe. ping
and wget both worked. But this needs a workaround in libngix app to avoid crash
```
diff --git a/src/core/nginx.c b/src/core/nginx.c
index 9fcb0eb..3afcb66 100644
@@ -1101,18 +1103,11 @@ ngx_core_module_init_conf(ngx_cycle_t *cycle, void 
*conf)

     if (ccf->user == (uid_t) NGX_CONF_UNSET_UINT && geteuid() == 0) {
         struct group   *grp;
-       struct passwd  *pwd;

         ngx_set_errno(0);
-       pwd = getpwnam(NGX_USER);
-       if (pwd == NULL) {
-           ngx_log_error(NGX_LOG_EMERG, cycle->log, ngx_errno,
-                         "getpwnam(\"" NGX_USER "\") failed");
-           return NGX_CONF_ERROR;
-       }

         ccf->username = NGX_USER;
-       ccf->user = pwd->pw_uid;
+       ccf->user = 0;

         ngx_set_errno(0);
         grp = getgrnam(NGX_GROUP);
@@ -1193,7 +1188,6 @@ ngx_set_user(ngx_conf_t *cf, ngx_command_t *cmd, void 
*conf)
     ngx_core_conf_t  *ccf = conf;

     char            *group;
-    struct passwd    *pwd;
     struct group     *grp;
     ngx_str_t  *value;

@@ -1214,14 +1208,8 @@ ngx_set_user(ngx_conf_t *cf, ngx_command_t *cmd, void 
*conf)

     ccf->username = (char *) value[1].data;

     ngx_set_errno(0);
-    pwd = getpwnam((const char *) value[1].data);
-    if (pwd == NULL) {
-       ngx_conf_log_error(NGX_LOG_EMERG, cf, ngx_errno,
-                          "getpwnam(\"%s\") failed", value[1].data);
-       return NGX_CONF_ERROR;
-    }

-    ccf->user = pwd->pw_uid;
+    ccf->user = 0;

     group = (char *) ((cf->args->nelts == 2) ? value[1].data : value[2].data);
```


Notes:
======
 - There might be many redundant macro definitions btw arm and x86 in
   pci_bus.h. I will fix it in the future.

justin-he committed:
plat/platform: Introduce new platform bus

platform bus is a simple bus interface for many devices, including
virtio_mmio device on arm64

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/virtio: Add new virtio_mmio device/driver on arm64

Based on [1], but device removing/releasing is not supported.

[1] http://docs.oasis-open.org/virtio/virtio/v1.0/cs04/virtio-v1.0-cs04.html#x1-1090002

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/virtio: Support virtio_cread/cwrite on arm64

Add virtio_cread/cwrite on arm64 with. Also add iowmb/iormb

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/virtio: Introduce several helpers for virtio ring

This patch provides several virtqueue helpers for virtio mmio
initialization.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/kvm: arm64: Fix arm64 memory layout for pcie ecam

In previous memory layout, it doesn't consider the device memory type
for pci ecam controller. This patch creates pci ecam device memory
started from 0x40,0000,0000 for pci config space on arm64.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
build: Make KVM_PCI default enable on arm64

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/pci_ecam: Introduce pci/pci_ecam.h

On arm64. we use a special generic pci controller, currently it
supports ECAM mode only on arm64 kvm plat.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/pci_bus: Change pci_device.base from u16 to unsigned long

On arm64, no ioport at all. It uses iomem instead, hence change the
base from u16 to unsigned long.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/virtio: Fix virtio net device mac negotiation

Otherwise, it causes virtio net mac negotiates failure on arm64.

Signed-off-by: Jia He <justin.he@arm.com>

justin-he committed:
plat/virtio: Change irq,base from u16 to u64 on arm64

On arm64, it only supports ECAM pci controller, hence u16 irq,base
is not enough.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/pci_ecam: Introduce pci ecam skeleton

This patch adds support for a generic PCI host controller. The controller
itself has no configuration registers and has its address spaces described
entirely by the device-tree. Currently, only ECAM is supported.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
lib/fdt: Fix fdt_address_cell and fdt_size_cell

The address-cell and size_cell should be probed from parent.
If the top parent doesn't provide it, return the default value.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/ofw: Export fdt_find_irq_parent_offset

This changes fdt_find_irq_parent_offset from static to extern for pci ecam
to use.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/ofw: Support completed ranges mapping

Previously, we only support 1:1 mapping for ranges property. This
is not enough for pci ecam on arm64.
e.g.
pcie@10000000 {
		interrupt-map-mask = < 0x1800 0x00 0x00 0x07 >;
		interrupt-map = < ... >;
		#interrupt-cells = < 0x01 >;
		ranges = < 0x1000000 0x00 0x00 0x00 0x3eff0000 0x00 0x10000 0x2000000 0x00 0x10000000 0x00 0x10000000 0x00 0x2eff0000 0x3000000 0x80 0x00 0x80 0x00 0x80 0x00 >;
		reg = < 0x40 0x10000000 0x00 0x10000000 >;
		msi-parent = < 0x8002 >;
		dma-coherent;
		bus-range = < 0x00 0xff >;
		linux,pci-domain = < 0x00 >;
		#size-cells = < 0x02 >;
		#address-cells = < 0x03 >;
		device_type = "pci";
		compatible = "pci-host-ecam-generic";
	};

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/ofw: Add fdt_node_offset_idx_by_compatible_list

The helper fdt_node_offset_idx_by_compatible_list is similar as
fdt_node_offset_by_compatible_list except that it returns the index
of compatible_list[] array when compatible string is matched.
This is useful when supporting multiple types of platform devices on
arm64.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/ofw: Add fdt_prop_read_bool

This fdt_prop_read_bool helper returns true when the property is
existed. This is helpful when we just need to know whether the property
is there instead of getting its value.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
lib/ukbus: Adjust default bus register priority

On arm64, platform bus should be probed before pci bus since the pci
controller is considered as a platform device. Thus set the default
priority from 1 to 2, then platform bus priority will be 1, higher
than pci bus.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
build: Enable pci library on arm64

This enables pci library on arm64

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
build: Enable the platform bus and virtio_mmio device on arm64

Enable the config for KVM_PF and add virtio mmio to the build system.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/platform_bus: Add probe/match interface for platform devices

There is no hardware ID for platform device on arm64 device-tree,
thus it introduces an ID match interface to identify the device id.
device id should be registered as a constant value.

Also it introduces probe interface to parse the necessary resource
(irq, base/size) in device-tree.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/virtio_mmio: Fix virtqueue size for virtio 9p device

For virtio 9p, the virtqueue size is 128 statically. Without this patch,
virtio 9p device can't be initialized in arm64 kvm plat.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/kvm/arm: Add a memory hole at the beginning 128M

Previously, arm64 kvm will setup 0x0-0x3fffffff as a device mmio area.
Hence user can even read/write 0x0 arbitrarily without panic.

After whis patch, we can prevent the memory corruption to some extent
since this set the beginning 128M as NONE attr.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/arm: Make sure fpsimd sysregs writing take effects

As per [1], we need to add isb() to make sure sysreg writing take
effects.

[1] https://github.com/freebsd/freebsd/blob/1b2bc99e/sys/arm64/arm64/vfp.c#L67

Signed-off-by: Jia He <justin.he@arm.com>

justin-he committed:
plat/virtio_9p: Remove excessive check for max virtqueue size

For qemu virtio mmio read with VIRTIO_MMIO_QUEUE_NUM_MAX, it will
unconditionally return VIRTQUEUE_MAX_SIZE(1024) unlike virtio pci.
But it can set the number of queue size when seting up the virtqueue.
See vm_setup_vq which will be fixed in coming patch.

So the excessive check in virtio 9p is not necessary.

Signed-off-by: Jia He <justin.he@arm.com>

justin-he committed:
plat/kvm/arm: Fix x19 corruption by interrupt handler

There is a crashing when testing nginx on arm64:
        [0x401015bc] trap_el1_sync+0x50/0x58
        [0x4010209c] el1_sync+0x9c/0x100
        [0x401169c0] _vprint+0x158/0x318
        [0x40116ce0] _uk_vprintk+0x64/0x74
        [0x40116d74] _uk_printk+0x84/0x90
        [0x4017f2e8] netif_alloc_rxpkts+0xd0/0x1f8
        [0x4010e5d0] virtio_netdev_rx_fillup+0x78/0x2a4
        [0x4010f298] virtio_netdev_recv+0x20c/0x400
        [0x4017eddc] uk_netdev_rx_one+0x2c8/0x2d0
        [0x4017f8a4] uknetdev_input+0x158/0x348
        [0x4013e5fc] _dispatcher+0x110/0x114
        [0x4011d478] uk_sched_thread_exit+0/0xbc
        [0x40218f44] uk_free+0x1c/0x28
        [0x40219030] free+0x18/0x20
        [0x4012a300] sys_write+0x17c/0x188
        [0x4012dcfc] pwritev+0x6c/0xd0
        [...]

Previously, x18,x19 are incorrectly restored by old _sp_. x18 can be
used/restore freely, but x19 is callee save register which can't be
corrupted.

Signed-off-by: Jia He <justin.he@arm.com>

justin-he committed:
plat/pci_ecam: Fix a lot compilation warnings in pci_ecam

This fixes the compilation warnings when introducing pci_ecam

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/virtio_mmio: Implement platform probe for virtio_mmio

This implements the probe interface for virtio_mmio device upon
platform bus.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/pci_bus: Split specific code into different files

There is some difference for pci controller on arm64 and x86. On arm64,
only pci ecam is supported. Hence split arch dependent codes into arch
specific directory. No functional changes in this patch, especially for
x86 codes.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/platform_bus,virtio_mmio: Avoid useless log storm

This is to suppress and adjust some log line level for platform bus and
virtio_mmio. E.g. ENODEV is normal when qemu doesn't create any virtio
mmio devices in the platform bus probing.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he committed:
plat/pci_bus: Pass memory allocation parameter to arch_pci_probe

On Arm, arch_pci_probe will use it when probing the pci device.

Thanks for fixing the memory allocation crash on X86 from Razvan Virtan.

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he commented:
/cc @skuenzer @felipehuici

justin-he commented:
@virtanrazvan
[updated on 2021-May-24]

All patches have been tagged with r-b from Razvan

nderjung commented:
Hey @justin-he! Thanks for submitting these patches via Github! :) Just a few notes on my end:

 * We are working to fix the `arm64-kvm/build` job in the checks below.  This is a [seperate issue](https://github.com/unikraft/kraft/issues/57) currently under investigation.  We hope to get this fixed so your PR can be checked fully by the CI; and,
 * Can you please remove the git trailer `GitHub-Pull-Request:` in your commits messages, this is added automatically by the CI.

Thanks! :)

justin-he committed:
plat/virtio: Fix the return value check for virtio_cread_bytes_many()

The return value of virtio_cread_bytes_many can be:
1) is -1 if something wrong
2) is len (>=0) if everything is fine

So the check condition in vpci_legacy_pci_config_get should be adjusted
if rc!=len

Otherwise, the initialization of virtio-9p device will be failed on
arm64 kvm plat.

Signed-off-by: Jia He <justin.he@arm.com>

justin-he commented:
>     * cked

Okay, thanks. I will remove the line

justin-he commented:
> 
> 
> > ```
> > * cked
> > ```
> 
> Okay, thanks. I will remove the line

Done

nderjung commented:
Thanks for updating the commit messages :)

It looks like this PR is failing against `x86_64-kvm`:

```
In file included from /tmp/build/80754af9/built/unikraft/plat/common/include/pci/pci_bus.h:70,
                 from /tmp/build/80754af9/built/unikraft/plat/common/pci_bus.c:57:
/tmp/build/80754af9/built/unikraft/plat/common/include/pci/pci_ecam.h:43:36: error: unknown type name 'fdt32_t'
   43 | extern int gen_pci_irq_parse(const fdt32_t *addr, struct fdt_phandle_args *out_irq);
      |                                    ^~~~~~~
/tmp/build/80754af9/built/unikraft/plat/common/include/pci/pci_ecam.h:43:58: warning: 'struct fdt_phandle_args' declared inside parameter list will not be visible outside of this definition or declaration
   43 | extern int gen_pci_irq_parse(const fdt32_t *addr, struct fdt_phandle_args *out_irq);
      |                                                          ^~~~~~~~~~~~~~~~
make[1]: *** [/tmp/build/80754af9/built/unikraft/support/build/Makefile.build:50: /tmp/build/80754af9/built/apps/helloworld/build/libkvmpci/pci_bus.common.o] Error 1
make: *** [Makefile:990: sub-make] Error 2
make: Entering directory '/tmp/build/80754af9/built/unikraft'
```

Please also can you double check the checkpatches for db9a16b, d9c1bb5 and 0a2d26f. Thanks!

justin-he commented:
> Thanks for updating the commit messages :)
> 
> It looks like this PR is failing against `x86_64-kvm`:
> 
> ```
> In file included from /tmp/build/80754af9/built/unikraft/plat/common/include/pci/pci_bus.h:70,
>                  from /tmp/build/80754af9/built/unikraft/plat/common/pci_bus.c:57:
> /tmp/build/80754af9/built/unikraft/plat/common/include/pci/pci_ecam.h:43:36: error: unknown type name 'fdt32_t'
>    43 | extern int gen_pci_irq_parse(const fdt32_t *addr, struct fdt_phandle_args *out_irq);
>       |                                    ^~~~~~~
> /tmp/build/80754af9/built/unikraft/plat/common/include/pci/pci_ecam.h:43:58: warning: 'struct fdt_phandle_args' declared inside parameter list will not be visible outside of this definition or declaration
>    43 | extern int gen_pci_irq_parse(const fdt32_t *addr, struct fdt_phandle_args *out_irq);
>       |                                                          ^~~~~~~~~~~~~~~~
> make[1]: *** [/tmp/build/80754af9/built/unikraft/support/build/Makefile.build:50: /tmp/build/80754af9/built/apps/helloworld/build/libkvmpci/pci_bus.common.o] Error 1
> make: *** [Makefile:990: sub-make] Error 2
> make: Entering directory '/tmp/build/80754af9/built/unikraft'
> ```
> 
> Please also can you double check the checkpatches for [db9a16b]> > (https://github.com/unikraft/unikraft/commit/db9a16bcbd61ce6f4778d47045a351fc5c58899c), [d9c1bb5](https://github.com/unikraft/unikraft/commit/d9c1bb5ca23432c79657004a35acb6657d15aea7) and [0a2d26f](https://github.com/unikraft/unikraft/commit/0a2d26fcad45fcaaacb6a418e3da95b00026d167). Thanks!

Thanks, I need to adjust the Arm pci header file a little bit, not having impact on X86

justin-he committed:
plat/pci_bus: arm64: Implement add/probe interfaces on arm64

This implements arch specific add/probe interfaces on arm64

Signed-off-by: Jia He <justin.he@arm.com>
Reviewed-by: Razvan Virtan <virtanrazvan@gmail.com>

justin-he commented:
> > Thanks for updating the commit messages :)
> > It looks like this PR is failing against `x86_64-kvm`:
> > ```
> > In file included from /tmp/build/80754af9/built/unikraft/plat/common/include/pci/pci_bus.h:70,
> >                  from /tmp/build/80754af9/built/unikraft/plat/common/pci_bus.c:57:
> > /tmp/build/80754af9/built/unikraft/plat/common/include/pci/pci_ecam.h:43:36: error: unknown type name 'fdt32_t'
> >    43 | extern int gen_pci_irq_parse(const fdt32_t *addr, struct fdt_phandle_args *out_irq);
> >       |                                    ^~~~~~~
> > /tmp/build/80754af9/built/unikraft/plat/common/include/pci/pci_ecam.h:43:58: warning: 'struct fdt_phandle_args' declared inside parameter list will not be visible outside of this definition or declaration
> >    43 | extern int gen_pci_irq_parse(const fdt32_t *addr, struct fdt_phandle_args *out_irq);
> >       |                                                          ^~~~~~~~~~~~~~~~
> > make[1]: *** [/tmp/build/80754af9/built/unikraft/support/build/Makefile.build:50: /tmp/build/80754af9/built/apps/helloworld/build/libkvmpci/pci_bus.common.o] Error 1
> > make: *** [Makefile:990: sub-make] Error 2
> > make: Entering directory '/tmp/build/80754af9/built/unikraft'
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > Please also can you double check the checkpatches for [[db9a16b](https://github.com/unikraft/unikraft/commit/db9a16bcbd61ce6f4778d47045a351fc5c58899c)]> > ([db9a16b](https://github.com/unikraft/unikraft/commit/db9a16bcbd61ce6f4778d47045a351fc5c58899c)), [d9c1bb5](https://github.com/unikraft/unikraft/commit/d9c1bb5ca23432c79657004a35acb6657d15aea7) and [0a2d26f](https://github.com/unikraft/unikraft/commit/0a2d26fcad45fcaaacb6a418e3da95b00026d167). Thanks!
> 
> Thanks, I need to adjust the Arm pci header file a little bit, not having impact on X86

Done, I tested the compilation on X86

razvanvirtan commented on plat/common/platform_bus.c:
```diff
@@ -82,104 +93,96 @@ static inline struct pf_driver *pf_find_driver(struct pf_device_id *id)
-		ret = pf_driver_add_device(drv, &devid, dev_base, dev_irq);
-	}
+		ret = pf_driver_add_device(drv, dev);
+		if (ret < 0) {
```
I think the condition here should be `ret < 0  && ret != -ENODEV` in order to avoid confusing error log.

justin-he replied:
Yes，it has been changed in https://github.com/unikraft/unikraft/commit/8edd8dfa0d51d9a1cc2759ed28f11ff35bd656f8

razvanvirtan replied:
That commit changes a similar condition in `pf_driver_add_device()`, at line 107, but not this one. Actually, I think we could keep the same condition here, but remove the error line (179), because the error message will be already generated by the `pf_driver_add_device` call, if needed.

razvanvirtan replied:
The problem is caused by a difference between https://github.com/unikraft/unikraft/commit/8edd8dfa0d51d9a1cc2759ed28f11ff35bd656f8 and this patch: https://patchwork.unikraft.org/patch/738335/
In  https://github.com/unikraft/unikraft/commit/8edd8dfa0d51d9a1cc2759ed28f11ff35bd656f, line 179 is not removed.

justin-he replied:
Yes, thanks for the reminder anyway. :-)

justin-he replied:
@razvanvirtan Hi, I am a little bit confused. Do you still think I need to change like:
```
in static int pf_probe(void)
{
...
 		ret = pf_driver_add_device(drv, dev);
 		if (ret < 0 && ret != -ENODEV) 
                          uk_free(pfh.a, dev);
...
```

justin-he replied:
I thought uk_free should be invoked even if ret == -ENODEV, what do you think of it?

razvanvirtan replied:
Yes, `uk_free` should remain even if `ret == -ENODEV`. Only removing the error message in that place is enough, so everything looks good now.

justin-he reviewed (commented):
I hadn't changed the patch as suggested by @razvanvirtan

razvanvirtan commented:
All good from my side.

Reviewed-by: Răzvan Vîrtan <virtanrazvan@gmail.com>

nderjung commented:
Hey, received this error during merge process:
```
  > Applying 0028-plat-arm-Make-sure-fpsimd-sysregs-writing-take-effec.patch...
  |  plat/common/include/arm/arm64/cpu.h |    6 ++++++
  |  1 file changed, 6 insertions(+)
  | error: patch failed: plat/common/include/arm/arm64/cpu.h:162
  | error: plat/common/include/arm/arm64/cpu.h: patch does not apply
  | hint: Use 'git am --show-current-patch' to see the failed patch
  | Applying: plat/arm: Make sure fpsimd sysregs writing take effects
  | Patch failed at 0001 plat/arm: Make sure fpsimd sysregs writing take effects
  | When you have resolved this problem, run "git am --continue".
  | If you prefer to skip this patch, run "git am --skip" instead.
  | To restore the original branch and stop patching, run "git am --abort".
```

razvanvirtan commented:
It seems to work on my local, if I'm trying to merge this PR into branch `staging`. Here is what I'm using:
```
git fetch origin pull/208/head && git checkout FETCH_HEAD
git checkout -b virtio_mmio
git checkout staging
git merge virtio_mmio
```
Do you have a diff output?

justin-he commented:
Hi, @nderjung I forcedly push again upon commit dedd3d4 , seems no conflicts.

justin-he commented:
Updated with dropping the patch "plat/arm/generic_timer: Fix assert in ns_to_ticks" in previous series since it is overlapped with 
https://github.com/unikraft/unikraft/pull/175/commits/3864d6466b8feed60ab66eefc1e5fe217b59172f
